<div class="alert alert-info">
<font size="5"><b>Комментарий ревьюера</b></font>

Привет Артур! Меня зовут Марат, и я буду твоим ревьюером. Спешу сообщить что все ключевые этапы в работе выполнены,  с задачей тебе удалось справиться. По поводу обращения - в IT сфере принято общаться на «ты» :) Но, если привычней на «вы», дай знать. Как ревьюера моя задача помочь тебе в развитии, дав хорошие советы. Я внимательно посмотрю твой код, ознакомлюсь с твоими выводами и оставлю комментарии. Где то могу предложить небольшие исправление в коде, но ненавязчиво. Где потребуются уточнения, я оставлю много наводящих вопросов. Они помогут тебя с поиском верного решения.

Все мои комментарии размечены по цветам, для лучшего восприятия. 
    
<div class="alert alert-success">Зеленым цветом и словом «Успех» отмечены особо удачные и элегантные решения, которыми ты можешь гордиться. </div>
        
<div class="alert alert-warning">Желтым и значком словом «Совет», помечены решения у которых есть альтернативные решения, более оптимальные. Ты можешь найти их сразу и доработать проект, или отложить это на потом, для будущих проектах. Проект будет принят и без их доработки. </div>
        
<div class="alert alert-danger"> Красным цветом и значком словом «Ошибка» помечу твои решения, на которые стоит обратить внимание прежде всего. После их доработки проект будет принят. </div>
        
Залог успеха - работа сообща, взаимное уважение и работа в диалоге. Поэтому, помечай свои ответные комментарии на мои реплики заметным цветом или курсивом, так мне будет легче их отслеживать. Пожалуйста, не изменяй и не удаляй мои комментарии. Все это поможет выполнить повторную проверку быстрей.

</div>



---
    
Для лучшей коммуникации будет лучше если ты будешь оставлять ответные комментарий (cвои комментарии к исправлениям, вопросы), так результат будет лучше.  А чтобы ревьюры их не теряли, лучше их подсвечивать цветом (можешь выбрать свой, а можешь использовать   тот который я предложил ниже). И чтобы не возникло путаницы лучше оставлять указание на номер версии комментария по итерация:  

1 итерация

<div class="alert alert-info"> <b>Комментарии студентаV1:</b> Привет Марат!.</div>


2 итерация

<div class="alert alert-info"> <b>Комментарии студентаV2:</b> Исправлено.... </div>
    
    
И лучше подсвечивать свои комментарии цветом, для этого достаточно в ячейку markdown добавить: 

```html
<div class="alert alert-info"> <b>Комментарии студентаV2:</b> Исправлено.... </div>    

```    
    
Хорошая коммуникация залог успеха! )
    
---

<div class="alert alert-info"> <b>Комментарии студента:</b> Привет, Марат! Сейчас попробую сделать свой проект лучше </div>

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюераV2</b></font>



Успех 👍:



Привет Артур )


 

# Развитие индустрии игр в начале XXI века

- Автор: Артур Карапетян
- Дата: 07.03.2026

### Цели и задачи проекта

Индустрия видеоигр прошла большой путь. В этом исследовании мы изучим период с 2000 по 2013 год включительно и ответим на вопросы:
- Как игры рассматриваемого периода оценивались критиками и игроками?
- Как выглядит топ игровых платформ по количеству вышедших на них играх?

### Описание данных

Данные содержат информацию о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр:
- `Name` — название игры.
- `Platform` — название платформы.
- `Year of Release` — год выпуска игры.
- `Genre` — жанр игры.
- `NA sales` — продажи в Северной Америке (в миллионах проданных копий).
- `EU sales` — продажи в Европе (в миллионах проданных копий).
- `JP sales` — продажи в Японии (в миллионах проданных копий).
- `Other sales` — продажи в других странах (в миллионах проданных копий).
- `Critic Score` — оценка критиков (от 0 до 100).
- `User Score` — оценка пользователей (от 0 до 10).
- `Rating` — рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.

### Содержимое проекта

1. [Загрузка данных и знакомство с ними](#link1)
2. [Проверка ошибок в данных и их предобработка](#link2)
- [Названия, или метки, столбцов датафрейма](#link2)
- [Типы данных](#link3)
- [Наличие пропусков в данных](#link4)
- [Явные и неявные дубликаты в данных](#link5)
3. [Фильтрация данных](#link6)
4. [Категоризация данных](#link7)
5. [Итоговый вывод](#link8)

---

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:


Вступление в работу очень важно, так человек, который смотрит твой проект (а возможно потенциальный работодатель или заказчик будет просматривать твой проект если ты его выложишь на Github) будет сразу введен в курс дела


    
<div class="alert alert-warning">



Совет 🤔:


 
 
    
- Ещё хочу заметить что вовсе не обязательно копировать введение из брифа (что конечно не нарушение, и копирование не делает твою работу хуже). Но если у тебя есть какие-то мысли по поводу проекта,  ты хочешь его сравнить с проектами которые проходила до этого,  или ты хочешь переписать бриф своими словами (чтобы он выглядел более естественно),  или хочешь продемонстрировать понимание проекта - то это только приветствуется. Это покажет потенциальному заказчику или работодателю, что ты понимаешь задачу и можешь её решить    
    
    
- Можно пойти дальше, внедрив интерактивные  ссылки. Это сделает проект более удобным для использования. Например в каком-то разделе ты ссылаешься на график из другого раздела, и для удобства, можешь добавить гиперссылку "график смотреть тут", и при нажатии происходит перемещение к нужному графику
    
	
	

А так оформляем гиперссылку

```html

[Ссылка номер один](#link1)

``` 
А вот так оформляем то место куда переместимся при нажатии на гиперссылку
    
    
```html

<a class='anchor' id="link1"></a>

Тут какой-то текст

``` 	

<div class="alert alert-info"> <b>Комментарии студента:</b> Решил оставить введение без изменений, но добавил интерактивные ссылки. Спасибо, что подсказал как их делать. </div>

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюераV2</b></font>



Успех 👍:



👍Пожалуйста


 

<a class='anchor' id="link1"></a>
## 1. Загрузка данных и знакомство с ними

- Загрузим необходимые библиотеки для анализа данных и данные из датасета `new_games.csv`. Затем выведем основную информацию о данных с помощью метода `info()`, первые и последние строки датафрейма.


In [1]:
# Импортируем библиотеку pandas
import pandas as pd

In [2]:
# Выгружаем данные в датафрейм
df = pd.read_csv('https://code.s3.yandex.net/datasets/new_games.csv')



<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:

 
- Артур, правильно что разведен импорт и загрузку данных, это разные     действия, которые в одну ячейке помещать не стоит


 

- Познакомимся с данными: выводим первые и последние 5 строк, а также результат метода `info()`.


In [3]:
# выведем первые 5 строк
df.head()

,Name,Platform,Year of Release,Genre,NA sales,EU sales,JP sales,Other sales,Critic Score,User Score,Rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E
1,Super Mario Bros.,NES,1985.0,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009.0,Sports,15.61,10.93,3.28,2.95,80.0,8,E
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN


In [4]:
# выведем последние 5 строк
df.tail()

,Name,Platform,Year of Release,Genre,NA sales,EU sales,JP sales,Other sales,Critic Score,User Score,Rating
16951,Samurai Warriors: Sanada Maru,PS3,2016.0,Action,0.00,0.0,0.01,0.0,NaN,NaN,NaN
16952,LMA Manager 2007,X360,2006.0,Sports,0.00,0.01,0.0,0.0,NaN,NaN,NaN
16953,Haitaka no Psychedelica,PSV,2016.0,Adventure,0.00,0.0,0.01,0.0,NaN,NaN,NaN
16954,Spirits & Spells,GBA,2003.0,Platform,0.01,0.0,0.0,0.0,NaN,NaN,NaN
16955,Winning Post 8 2016,PSV,2016.0,Simulation,0.00,0.0,0.01,0.0,NaN,NaN,NaN


In [5]:
# Ознакомимся с общей информацией о датасете
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


In [6]:
# Выведем и сохраним изначальное количество строк
rows = df.shape[0]
rows

16956

Датасет содержит 16956 строк и 11 столбцов, в которых представлена информация о продажах видеоигр в различных регионах и их оценках от критиков и игроков.
В шести столбцах присутствуют пропуски.
Названия столбцов верно передают их суть, но в дальнейшем мы преобразуем их в snake_case формат.
Изучим типы данных и их корректность:
- **Числовые значения с плавающей запятой (float64).** Четыре столбца представлены типом `float64`.
    - Для столбцов `NA sales` и `Other sales` этот тип подходит, так как в них содержатся данные о продажах в миллионах проданных копий.
    - Столбец `Year of Release` содержит год выхода игры и его следует преобразовать в тип `int16`.
    - Столбец `Critic Score` содержит информацию об оценках критиков (применяется стобальная система оценок) и его следует преобразовать в тип `int8`.
    
- **Строковые данные (object).** Семь столбцов имеют тип данных `object`:
    - `Name`, `Platform` и `Genre` содержат строковую информацию (название игры, игровая платформа и жанр), что логично для текстовых данных. Здесь тип данных `object` подходит.
    - `Rating` также хранит текстовые данные о присвоенном возрастном рейтинге, но их можно рассматривать как категориальные признаки. В этом случае можно использовать тип `category`, чтобы улучшить производительность и оптимизировать память, если набор значений ограничен и заведомо известен.
    - `EU sales` и `JP sales` хранят информацию о продажах. Для таких данных необходимо использовать тип `float64`.
    - `User score` хранит информацию о пользовательских оценках. Так как в данном столбце встречаются дробные значения, для таких данных подойдёт тип `float64`.
    

После анализа типов данных видно, что для 6 столбцов необходимо преобразовать тип данных.

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:

 
- Знакомства с данными проведено, джентльменский набор info + head/tail/sample на месте


- Отличный анализ, нестыковки указаны,  план дальнейших действий составлен


 


---
<a class='anchor' id="link2"></a>
## 2.  Проверка ошибок в данных и их предобработка


### 2.1. Названия, или метки, столбцов датафрейма

- Выведем на экран названия всех столбцов датафрейма и проверим их стиль написания.
- Приведем все столбцы к стилю snake case в нижнем регистре.

In [7]:
# Выведем на экран названия всех столбцов
df.columns

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')

In [8]:
# Приведем названия столбцов к стилю snake case с помощью метода rename
df = df.rename(columns={'Name': 'name', 'Platform': 'platform', 'Year of Release': 'year_of_release', 'Genre': 'genre', 'NA sales': 'na_sales', 'EU sales': 'eu_sales', 'JP sales': 'jp_sales', 'Other sales': 'other_sales', 'Critic Score': 'critic_score', 'User Score': 'user_score', 'Rating': 'rating'})

In [9]:
# Проверим новые названия столбцов
df.columns

Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'],
      dtype='object')

 
	
<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюера</b></font>


Совет: 


  

- по условию задачи нам требуется не переименовать столбцы, а лишь привести их к нижнему регистру. Знаешь ли ты способ автоматического приведения всех столбов таблицы к строчному виду?! Это может быть полезно, когда у тебя не несколько столбцов, а сотни, что в жизни datascientists встречается довольно часто
    
 
     

<div class="alert alert-info"> <b>Комментарии студента:</b> Мне не знаком такой способ. Могу ошибаться, но в курсе он не описан, а на изучение дополнительных материалов пока не хватает времени.</div>


<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюераV2</b></font>



Совет 🤔:


 

 
    
- Тут можно было использовать .lower  +  .replace
    
     


- Есть и другие способы автоматизированного приведения к змеиному виду, например [регуляризация](https://habr.com/ru/articles/349860/)   Она пригодится если .lower() / .replace окажется недостаточным. Например если названия столбцов будут в верблюжьем (CamelCase) стиле, с помощью регуляризации легко перевести из верблюжьего в змеиный  
     
     
     df.columns = [re.sub(r'(?<!^)(?=[A-Z])', '_', i).lower() for i in df.columns]
     
     
     
     
     
       
   




<a class='anchor' id="link3"></a>
### 2.2. Типы данных

Изучим типы данных и их корректность

In [10]:
# Выведем на экран столбцы и их типы данных
df.dtypes

name                object
platform            object
year_of_release    float64
genre               object
na_sales           float64
eu_sales            object
jp_sales            object
other_sales        float64
critic_score       float64
user_score          object
rating              object
dtype: object

In [11]:
# Выведем уникальные значения столбцов, которые могут содержать не подходящий тип данных
display('Уникальные значения столбца year_of_release', df['year_of_release'].unique())
display('Уникальные значения столбца critic_score', df['critic_score'].unique())
display('Уникальные значения столбца eu_sales', df['eu_sales'].unique())
display('Уникальные значения столбца jp_sales', df['jp_sales'].unique())
display('Уникальные значения столбца user_score', df['user_score'].unique())

'Уникальные значения столбца year_of_release'

array([2006., 1985., 2008., 2009., 1996., 1989., 1984., 2005., 1999.,
       2007., 2010., 2013., 2004., 1990., 1988., 2002., 2001., 2011.,
       1998., 2015., 2012., 2014., 1992., 1997., 1993., 1994., 1982.,
       2016., 2003., 1986., 2000.,   nan, 1995., 1991., 1981., 1987.,
       1980., 1983.])

'Уникальные значения столбца critic_score'

array([76., nan, 82., 80., 89., 58., 87., 91., 61., 97., 95., 77., 88.,
       83., 94., 93., 85., 86., 98., 96., 90., 84., 73., 74., 78., 92.,
       71., 72., 68., 62., 49., 67., 81., 66., 56., 79., 70., 59., 64.,
       75., 60., 63., 69., 50., 25., 42., 44., 55., 48., 57., 29., 47.,
       65., 54., 20., 53., 37., 38., 33., 52., 30., 32., 43., 45., 51.,
       40., 46., 39., 34., 35., 41., 36., 28., 31., 27., 26., 19., 23.,
       24., 21., 17., 22., 13.])

'Уникальные значения столбца eu_sales'

array(['28.96', '3.58', '12.76', '10.93', '8.89', '2.26', '9.14', '9.18',
       '6.94', '0.63', '10.95', '7.47', '6.18', '8.03', '4.89', '8.49',
       '9.09', '0.4', '3.75', '9.2', '4.46', '2.71', '3.44', '5.14',
       '5.49', '3.9', '5.35', '3.17', '5.09', '4.24', '5.04', '5.86',
       '3.68', '4.19', '5.73', '3.59', '4.51', '2.55', '4.02', '4.37',
       '6.31', '3.45', '2.81', '2.85', '3.49', '0.01', '3.35', '2.04',
       '3.07', '3.87', '3.0', '4.82', '3.64', '2.15', '3.69', '2.65',
       '2.56', '3.11', '3.14', '1.94', '1.95', '2.47', '2.28', '3.42',
       '3.63', '2.36', '1.71', '1.85', '2.79', '1.24', '6.12', '1.53',
       '3.47', '2.24', '5.01', '2.01', '1.72', '2.07', '6.42', '3.86',
       '0.45', '3.48', '1.89', '5.75', '2.17', '1.37', '2.35', '1.18',
       '2.11', '1.88', '2.83', '2.99', '2.89', '3.27', '2.22', '2.14',
       '1.45', '1.75', '1.04', '1.77', '3.02', '2.75', '2.16', '1.9',
       '2.59', '2.2', '4.3', '0.93', '2.53', '2.52', '1.79', '1.3', '2.6',
   

'Уникальные значения столбца jp_sales'

array(['3.77', '6.81', '3.79', '3.28', '10.22', '4.22', '6.5', '2.93',
       '4.7', '0.28', '1.93', '4.13', '7.2', '3.6', '0.24', '2.53',
       '0.98', '0.41', '3.54', '4.16', '6.04', '4.18', '3.84', '0.06',
       '0.47', '5.38', '5.32', '5.65', '1.87', '0.13', '3.12', '0.36',
       '0.11', '4.35', '0.65', '0.07', '0.08', '0.49', '0.3', '2.66',
       '2.69', '0.48', '0.38', '5.33', '1.91', '3.96', '3.1', '1.1',
       '1.2', '0.14', '2.54', '2.14', '0.81', '2.12', '0.44', '3.15',
       '1.25', '0.04', '0.0', '2.47', '2.23', '1.69', '0.01', '3.0',
       '0.02', '4.39', '1.98', '0.1', '3.81', '0.05', '2.49', '1.58',
       '3.14', '2.73', '0.66', '0.22', '3.63', '1.45', '1.31', '2.43',
       '0.7', '0.35', '1.4', '0.6', '2.26', '1.42', '1.28', '1.39',
       '0.87', '0.17', '0.94', '0.19', '0.21', '1.6', '0.16', '1.03',
       '0.25', '2.06', '1.49', '1.29', '0.09', '2.87', '0.03', '0.78',
       '0.83', '2.33', '2.02', '1.36', '1.81', '1.97', '0.91', '0.99',
       '0.95', '2.0'

'Уникальные значения столбца user_score'

array(['8', nan, '8.3', '8.5', '6.6', '8.4', '8.6', '7.7', '6.3', '7.4',
       '8.2', '9', '7.9', '8.1', '8.7', '7.1', '3.4', '5.3', '4.8', '3.2',
       '8.9', '6.4', '7.8', '7.5', '2.6', '7.2', '9.2', '7', '7.3', '4.3',
       '7.6', '5.7', '5', '9.1', '6.5', 'tbd', '8.8', '6.9', '9.4', '6.8',
       '6.1', '6.7', '5.4', '4', '4.9', '4.5', '9.3', '6.2', '4.2', '6',
       '3.7', '4.1', '5.8', '5.6', '5.5', '4.4', '4.6', '5.9', '3.9',
       '3.1', '2.9', '5.2', '3.3', '4.7', '5.1', '3.5', '2.5', '1.9', '3',
       '2.7', '2.2', '2', '9.5', '2.1', '3.6', '2.8', '1.8', '3.8', '0',
       '1.6', '9.6', '2.4', '1.7', '1.1', '0.3', '1.5', '0.7', '1.2',
       '2.3', '0.5', '1.3', '0.2', '0.6', '1.4', '0.9', '1', '9.7'],
      dtype=object)

Датасет содержит следующие типы данных:
- **Числовые значения с плавающей запятой (float64).** Четыре столбца представлены типом `float64`.
    - Для столбцов `na_sales` и `other_sales` этот тип подходит, так как в них содержатся данные о продажах в миллионах проданных копий.
    - Столбец `year_of_release` содержит год выхода игры и его следует преобразовать в тип `int64`.
    - Столбец `critic_Score` содержит информацию об оценках критиков (применяется стобальная система оценок). Эти данные мы преобразуем в тип `int64` после обработки пропусков.
- **Строковые данные (object).** Семь столбцов имеют тип данных `object`:
    - `name`, `platform`, `genre` и `rating` содержат строковую информацию (название игры, игровая платформа и жанр), что логично для текстовых данных. Здесь тип данных `object` подходит.
    - `eu_sales` и `jp_sales` хранят информацию о продажах. Для таких данных необходимо использовать тип `float64`.
    - `user_score` хранит информацию о пользовательских оценках. Эти данные мы преобразуем в тип `int64` после обработки пропусков.
    

Анализ типов данных показал, что для 6 столбцов необходимо преобразовать тип данных.


    
<div class="alert alert-danger">
<font size="5"><b>Комментарий ревьюера</b></font>



Ошибка ❌:


  
    
Верно, но ещё стоило   вывести уникальные значения по   столбцам где мы видим несоответствия между типом данных и названием столбцов, и проанализировать увиденное

 
  
Например, если у нас в числовом столбце присутствует текст 'unknown', тут всё понятно, но кое-где встречается слово "tbd",  возможно   оно означает нечто иное, и из него не стоит делать пропуск,  а можно сделать цифру    
 

<div class="alert alert-info"> <b>Комментарии студента:</b> Добавил ячейку, где я вывожу уникальные значения тех столбцов, в которых считаю тип данных неподходящим. В столбце user_score действительно встретилось значение "tbd". Эта аббревиатура "to be deternined", что означает "подлежит определению". Думаю, в контексте наших данных допустимо считать такое значение эквивалентным пропуску.</div>

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюераV2</b></font>



Успех 👍:



👍Согласен


 

Преобразуем столбцы с числовыми значениями в числовые типы данных. При этом данные, которые не удастся преобразовать заменим на NaN.

In [12]:
# Преобразуем столбцы с числовыми значениями в числовые типы данных
for column in ['year_of_release', 'critic_score', 'eu_sales', 'jp_sales', 'user_score']:
    df[column] = pd.to_numeric(df[column], errors='coerce')

Столбец `year_of_release` содержит целые числа, поэтому преобразуем его в целочисленный формат.

In [13]:
# Преобразуем столбец 'year_of_release' в целочисленный формат
df['year_of_release'] = df['year_of_release'].astype('Int64')

Проверим результат преобразований. Снова выведем на экран столбцы и их типы данных.

In [14]:
# Выведем на экран столбцы и их типы данных
df.dtypes

name                object
platform            object
year_of_release      Int64
genre               object
na_sales           float64
eu_sales           float64
jp_sales           float64
other_sales        float64
critic_score       float64
user_score         float64
rating              object
dtype: object

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>



Успех 👍:



    
- Плюс за проверку      
    
    
    
    .to_numeric( , errors='coerce')


- Отлично, всё как просят в проекте
    
    

     
- 👍 'Int64' 

<div class="alert alert-warning">



Совет 🤔:

 

 

- Можно сразу привезти к целочисленному 	 и critic_score. Один из плюсов - это экономия ресурсов, ведь int, занимает гораздо меньше памяти чем float 
 

<div class="alert alert-info"> <b>Комментарии студента:</b> Изначально столбцы critic_score и user_score я сразу хотел привести к типу int, но преобразование не проходило из-за наличия пропусков. Эти два столбца я преобразовываю в int в следующем понкте "Наличие пропусков в данных".</div>

<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюераV2</b></font>



Совет 🤔:

user_score у нас float. 
    
    
    о преобразование не проходило из-за наличия пропусков. Эти два столбца я преобразовываю в int в следующем понкте "Наличие пропусков в данных".
    
    
Ты же выше используешь    .astype('Int64'), он работает даже если в столбце пропуски 

Для последующей обработки и анализа, данные преобразованы в подходящие по характеру типы данных.

<a class='anchor' id="link4"></a>
### 2.3. Наличие пропусков в данных

- Посчитаем количество пропусков в каждом столбце в абсолютных и относительных значениях.


In [15]:
# Посчитаем количество пропусков в каждом столбце с помощью метода isna()
df.isna().sum()

name                  2
platform              0
year_of_release     275
genre                 2
na_sales              0
eu_sales              6
jp_sales              4
other_sales           0
critic_score       8714
user_score         9268
rating             6871
dtype: int64

In [16]:
# Посчитаем долю пропусков в каждом столбце
df.isna().sum() / len(df) * 100

name                0.011795
platform            0.000000
year_of_release     1.621845
genre               0.011795
na_sales            0.000000
eu_sales            0.035386
jp_sales            0.023590
other_sales         0.000000
critic_score       51.391838
user_score         54.659118
rating             40.522529
dtype: float64

Изучим данные с пропущенными значениями:

- Столбцы `name`, `genre`, `eu_sales` и `jp_sales` имеют долю пропущенных значений менее 1%. Пропуски в столбцах `name` и `genre` могли произойти по техническим причинам или из-за человеческого фактора. Иными словами их можно отнести категории полностью случайных пропусков (**MCAR**). Пропуски в столбцах `eu_sales` и `jp_sales` могут быть следствием малого количества проданных копий игр или отсутвием данных о продажах. Такие пропуски относятся к категории случайных пропусков, которые объясняются наблюдаемыми данными или значениями (**MAR**). Данные пропуски ***заменим на среднее значение в зависимости от названия платформы и года выхода игры***.

- Столбец `year_of_release` имеет 275 пропусков, что составляет около 2% общего количества значений. Данные в этом столбце имеют важное значение, так как в предстоящем исследовании потребуется фильтрация по году выпуска. Причины наличия пропусков сложно определить, поэтому отнесем эти пропуски к категории пропусков связаных с факторами, которых нет в наборе данных (**MNAR**). Пропуски в этих данных мы будем ***заменим на значения-индикаторы***.

- Столбцы `critic_score`, `user_score` и `rating` имеют доли пропусков 51%, 55% и 41% соответственно. Пропуски в этих столбцах можно объснить характером данных. Игра могла выйти давно и систем оценок и рейтингов тогда не существовало. Также вероятно, что релиз игры прошел незамеченным, не имел финансового успеха и в итоге остался не оцененным. Данные пропуски относятся к категории пропусков, которые объясняются наблюдаемыми данными (**MAR**). Для целей исследования столбцы `critic_score` и `user_score` имеют важное значение, поэтому эти пропуски мы ***заполним нулевыми значениями***. Данные столбца `rating` в исследовании учитываться не будут, поэтому в этом столбце пропуски мы будем ***игнорировать***.


<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>



Успех 👍:


    
- Расписаны причины возможных пропусков 
    


 
    
    
<div class="alert alert-warning">


Совет 🤔:

       
  
 



- А может   дело в страновых рейтингах? Например японская игрушка может не получить рейтинг США. То что тоже можно поизучать вопрос    

<div class="alert alert-info"> <b>Комментарии студента:</b> Это действительно могло быть причиной наличия пропусков. Честно говоря, о рейтинге ESRB и о том, как он действует в разных странах, я не стал искать дополнительной информации, так как для исследования эти данные (столбец rating) мне показались не важными.</div>

In [17]:
# Удалим пропуски в столбцах name, genre, eu_sales и jp_sales
df = df.dropna(subset=['name', 'genre', 'eu_sales', 'jp_sales'])

In [18]:
# Заменим пропуски в столбце eu_sales на среднее значение в зависимости от названия платформы и года выхода игры
def mean_group_eu_sales(row):
    group = df[(df['platform'] == row['platform']) & ((df['year_of_release'] == row['year_of_release']))]
    return group['eu_sales'].mean()
df.apply(mean_group_eu_sales, axis=1)

0        0.991364
1        0.425455
2        0.168556
3        0.178489
4        1.561667
           ...   
16951    0.032564
16952    0.109785
16953    0.004535
16954    0.086903
16955    0.004535
Length: 16944, dtype: float64

In [19]:
# Заменим пропуски в столбце jp_sales на среднее значение в зависимости от названия платформы и года выхода игры
def mean_group_jp_sales(row):
    group = df[(df['platform'] == row['platform']) & ((df['year_of_release'] == row['year_of_release']))]
    return group['jp_sales'].mean()
df.apply(mean_group_jp_sales, axis=1)

0        0.207955
1        1.321818
2        0.044683
3        0.050937
4        2.151667
           ...   
16951    0.033333
16952    0.012043
16953    0.041512
16954    0.061613
16955    0.041512
Length: 16944, dtype: float64



<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>



Успех 👍:



- Здорово что попробовал заполнить пропуски в группировке


 

<div class="alert alert-warning">


Совет 🤔:

     
   
    
И тоже   самое можно сделать проще, нагляднее (меньше кода), эффективней (использование векторных операций), с помощью  [transform](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.transform.html#pandas-dataframe-transform)
    
	
		df['jp_sales'] = df['jp_sales'].fillna(df.groupby(['platform', 'year_of_release'])['jp_sales'].transform('mean'))
    
    
Как видишь в отличие от  apply, для него даже не придётся писать функцию, потому что он умеет работать с   встроенныеми функциямикак mean, sum, max, min      
    
     

<div class="alert alert-info"> <b>Комментарии студента:</b> Действительно отличный способ обойтись без функции. Возьму на заметку. Большое спасибо!</div>

In [20]:
# Заменим пропуски в столбце year_of_release на значения-индикатор
df['year_of_release'] = df['year_of_release'].fillna(-1)

In [21]:
# Заполним пропуски в столбцах critic_score и user_score нулевыми значениями
df['critic_score'] = df['critic_score'].fillna(-1)
df['user_score'] = df['user_score'].fillna(-1)

   
    
<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюера</b></font>


Совет 🤔:


    
    
- Качество заглушки лучше выбрать явно невозможное число, обычно это  -1 
    
    
- Был вариант не использовать заглушку. Минус заглушки в том что с ней    мы должны держать постоянно в голове что данные не реальные  и   при проведении анализа, подсчете средних, создании сводных таблиц с агрегированными показателями итп итд мы должны будем   постоянно их убирать,  и в какой-то момент можем просто забыть сделать это (вот поэтому кстати и стоит использовать "видную" заглушку, например -1).  А наличие пропусков в столбце   нисколько не помешает посчитать корректно ту же среднюю, построить сводную таблицу итд
    
    
    
    

<div class="alert alert-info"> <b>Комментарии студента:</b> Внес исправления. Теперь заглушки это -1</div>

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюераV2</b></font>



Успех 👍:



👍


(Хотя по-моему лучше вообще убрать) 

Теперь когда в столбцах `critic_score` и `user_score` все значения являются числовыми, для удобства их можно преобразовать в формат `int64`

In [22]:
# Преобразуем значения critic_score и user_score в int64
df['critic_score'] = df['critic_score'].astype('int64')
df['user_score'] = df['user_score'].astype('int64')

In [23]:
# Проверим результат и снова выведем количество пропусков в каждом столбце с помощью метода isna()
df.isna().sum()

name                  0
platform              0
year_of_release       0
genre                 0
na_sales              0
eu_sales              0
jp_sales              0
other_sales           0
critic_score          0
user_score            0
rating             6868
dtype: int64

In [24]:
# Проверим количество строк после обработки пропусков
df.shape[0]

16944

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>



Успех 👍:

 
 
    
- Решение по работе с пропусками обосновано 
    
    

    

<a class='anchor' id="link5"></a>
### 2.4. Явные и неявные дубликаты в данных

- Изучим уникальные значения в категориальных данных с названиями жанра игры, платформы, рейтинга и года выпуска. Проверим, встречаются ли среди данных неявные дубликаты, связанные с опечатками или разным способом написания.
- При необходимости проведите нормализацию данных с текстовыми значениями. Названия или жанры игр можно привести к нижнему регистру, а названия рейтинга — к верхнему.

In [25]:
# Выведем уникальные значения в столбце genre
df['genre'].unique()

array(['Sports', 'Platform', 'Racing', 'Role-Playing', 'Puzzle', 'Misc',
       'Shooter', 'Simulation', 'Action', 'Fighting', 'Adventure',
       'Strategy', 'MISC', 'ROLE-PLAYING', 'RACING', 'ACTION', 'SHOOTER',
       'FIGHTING', 'SPORTS', 'PLATFORM', 'ADVENTURE', 'SIMULATION',
       'PUZZLE', 'STRATEGY'], dtype=object)

In [26]:
# Выведем уникальные значения в столбце platform
df['platform'].unique()

array(['Wii', 'NES', 'GB', 'DS', 'X360', 'PS3', 'PS2', 'SNES', 'GBA',
       'PS4', '3DS', 'N64', 'PS', 'XB', 'PC', '2600', 'PSP', 'XOne',
       'WiiU', 'GC', 'GEN', 'DC', 'PSV', 'SAT', 'SCD', 'WS', 'NG', 'TG16',
       '3DO', 'GG', 'PCFX'], dtype=object)

In [27]:
# Выведем уникальные значения в столбце rating
df['rating'].unique()

array(['E', nan, 'M', 'T', 'E10+', 'K-A', 'AO', 'EC', 'RP'], dtype=object)

In [28]:
# Выведем уникальные значения в столбце year_of_release
df['year_of_release'].unique()

<IntegerArray>
[2006, 1985, 2008, 2009, 1996, 1989, 1984, 2005, 1999, 2007, 2010, 2013, 2004,
 1990, 1988, 2002, 2001, 2011, 1998, 2015, 2012, 2014, 1992, 1997, 1993, 1994,
 1982, 2016, 2003, 1986, 2000,   -1, 1995, 1991, 1981, 1987, 1980, 1983]
Length: 38, dtype: Int64

- Столбец `genre` содержит неявные дубликаты. Преобразуем значения этого столбца в нижний регистр.

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>



Успех 👍:



👍Проверил,  проанализировал, написал как будешь исправлять.  Всё как надо


 

In [29]:
# Преобразуем значения столбца genre в нижний регистр
df['genre'] = df['genre'].str.lower()

In [30]:
# Проверим результат преобразования
df['genre'].unique()

array(['sports', 'platform', 'racing', 'role-playing', 'puzzle', 'misc',
       'shooter', 'simulation', 'action', 'fighting', 'adventure',
       'strategy'], dtype=object)

- Проверим наличие явных дубликатов в данных.

In [31]:
# Выводим количество явных дубликатов
df.duplicated().sum()

241

In [32]:
# Удалим явные дубликаты
df = df.drop_duplicates()

In [33]:
# Проверим количество явных дубликатов после очистки
df.duplicated().sum()

0

In [34]:
# Проверим количество строк после обработки пропусков и дубликатов
rows_cleaned = df.shape[0]
rows_cleaned

16703

- В результате обработки категориальных данных были обнаружены дублирующие значения в столбце `genre`, написанные в разных стилях. Все значения в этих данных были приведены к общему стилю, после чего было обнаружено 241 явных дубликатов, которые были удалены из датасета.

- Посчитаем количество удалённых строк в абсолютном и относительном значениях.

In [35]:
# Посчитаем количество удаленных строк
deleted_rows = rows - rows_cleaned
deleted_rows

253

In [36]:
# Посчитаем долю удаленных строк
deleted_rows_ratio = round((deleted_rows / rows * 100), 1)
deleted_rows_ratio

1.5

**В результате предобработки были проведены операции по подготовке датасета к анализу:**
 - Преобразование названий столбцов
 - Преобразование типов данных
 - Обработка пропущенных значений
 - Обработка неявных дубликатов и удаление явных дубликатов
 
 В ходе очистки данных было удалено 253 строки, что составляет 1,5% от изначального количества строк.


<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>



Успех 👍:



- Обращено внимание на скрытые и явные дубликаты 


- Здорово что предварительно просмотрены уникальные значения 
 
 
- Есть контроль на сколько уменьшилась выборка 
    
    
    
     
    
    
<div class="alert alert-warning">
 


Совет 🤔:
 

 
- Что можно было сделать еще?

Поискать неявные дубликаты   в рейтингах и названиях игр. Данные стали бы чуть чуть чище. 


На примере рейтингов это могло выглядеть так: 

1. Используем value_counts() к rating и обнаруживаем небольшое число у K-A и EC 
2. Гуглим и обнаруживаем что K-A это старая версия E (можно продолжить выяснять).. 
3. Делаем замены. Такой вариант работы с неявными дубликатами. 
 

<div class="alert alert-info"> <b>Комментарии студента:</b> Если искать неявные дубликаты в названиях игр через вывод уникальных значений, то процесс поиска неявных дубликатов был бы достаточно трудоемким, ведь строк в датасете почти 17000. Возможно я просто не знаю другого способа или неправильно понимаю твое замечание. А рейтинг ESRB для исследования не играет роли. Допускаю, что я могу ошибаться</div>

<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюераV2</b></font>



Совет 🤔:


 Действительно с играми не просто,  скорее всего тут надо будет использовать какие-то специальные инструменты

---
<a class='anchor' id="link6"></a>
## 3. Фильтрация данных

Цель исследования - изучение истории продаж игр в начале XXI века, а именно период с 2000 по 2013 год включительно. Отберем данные по этому показателю. Сохраним новый срез данных в отдельном датафрейме `df_actual`.

In [37]:
# Отберем данные за период с 2000 по 2013 год включительно и сохраним новый срез данных в отдельном датафрейме df_actual
df_actual = df[df['year_of_release'].between(2000, 2013)].copy()



<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>



Успех 👍:



👍
 

In [38]:
# Выведем получившийся датасет
df_actual

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
0,Wii Sports,Wii,2006,sports,41.36,28.96,3.77,8.45,76,8,E
2,Mario Kart Wii,Wii,2008,racing,15.68,12.76,3.79,3.29,82,8,E
3,Wii Sports Resort,Wii,2009,sports,15.61,10.93,3.28,2.95,80,8,E
6,New Super Mario Bros.,DS,2006,platform,11.28,9.14,6.50,2.88,89,8,E
7,Wii Play,Wii,2006,misc,13.96,9.18,2.93,2.84,58,6,E
...,...,...,...,...,...,...,...,...,...,...,...
16947,Men in Black II: Alien Escape,GC,2003,shooter,0.01,0.00,0.00,0.00,-1,-1,T
16949,Woody Woodpecker in Crazy Castle 5,GBA,2002,platform,0.01,0.00,0.00,0.00,-1,-1,NaN
16950,SCORE International Baja 1000: The Official Game,PS2,2008,racing,0.00,0.00,0.00,0.00,-1,-1,NaN
16952,LMA Manager 2007,X360,2006,sports,0.00,0.01,0.00,0.00,-1,-1,NaN


In [39]:
# Ознакомимся с общей информацией о новом датасете
df_actual.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12772 entries, 0 to 16954
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             12772 non-null  object 
 1   platform         12772 non-null  object 
 2   year_of_release  12772 non-null  Int64  
 3   genre            12772 non-null  object 
 4   na_sales         12772 non-null  float64
 5   eu_sales         12772 non-null  float64
 6   jp_sales         12772 non-null  float64
 7   other_sales      12772 non-null  float64
 8   critic_score     12772 non-null  int64  
 9   user_score       12772 non-null  int64  
 10  rating           8715 non-null   object 
dtypes: Int64(1), float64(4), int64(2), object(4)
memory usage: 1.2+ MB


В результате фильтрации по необходимым для анализа годам, мы получили срез данных, состоящий из 12772 строк.

---
<a class='anchor' id="link7"></a>
## 4. Категоризация данных
    
Проведем категоризацию данных:
- Разделим все игры по оценкам пользователей и выделите такие категории: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).

In [40]:
# Создадим функцию для категоризации оценок и добавим в датасет новый столбец, в котором оценки пользователей будут распределяться по категориям
def categorize_user_score(score):
    if 8 <= score <= 10:
        return 'high score'
    elif 3 <= score < 8:
        return 'middle score'
    elif 0 <= score < 3:
        return 'low score'
    else:
        return 'not rated'
df_actual['user_categories'] = df_actual['user_score'].apply(categorize_user_score)

 
 

<div class="alert alert-danger">
<font size="5"><b>Комментарий ревьюера</b></font>



Ошибка ❌:

Функция правильная,  но предварительно вытаскиваем заглушку (Или учитываем в функции что 0 это отсутствие данныых а не score)
 

<div class="alert alert-info"> <b>Комментарии студента:</b> Так как в critic_score и user_score ранее я поменял заглушки на -1, то в категоризацию я добавил дополнительную категорию "not rated" для таких данных</div>

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюераV2</b></font>



Успех 👍:



👍Как вариант


 

In [41]:
# Переместим новый столбец на место рядом со столбцом 'critic_score'
user_categories = df_actual.pop('user_categories')
df_actual.insert(10, 'user_categories', user_categories)

In [42]:
# Проверим результат
df_actual.head()

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,user_categories,rating
0,Wii Sports,Wii,2006,sports,41.36,28.96,3.77,8.45,76,8,high score,E
2,Mario Kart Wii,Wii,2008,racing,15.68,12.76,3.79,3.29,82,8,high score,E
3,Wii Sports Resort,Wii,2009,sports,15.61,10.93,3.28,2.95,80,8,high score,E
6,New Super Mario Bros.,DS,2006,platform,11.28,9.14,6.50,2.88,89,8,high score,E
7,Wii Play,Wii,2006,misc,13.96,9.18,2.93,2.84,58,6,middle score,E


- Разделим все игры по оценкам критиков и выделите такие категории: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).

In [43]:
# Создадим функцию для категоризации оценок и добавим в датасет новый столбец, в котором оценки критиков будут распределяться по категориям
def categorize_critic_score(score):
    if 80 <= score <= 100:
        return 'high score'
    elif 30 <= score < 80:
        return 'middle score'
    elif 0 <= score < 3:
        return 'low score'
    else:
        return 'not rated'
df_actual['critic_categories'] = df_actual['critic_score'].apply(categorize_critic_score)

In [44]:
# Переместим новый столбец на место рядом со столбцом 'critic_score'
critic_categories = df_actual.pop('critic_categories')
df_actual.insert(9, 'critic_categories', critic_categories)

In [45]:
# Проверим результат
df_actual.head()

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,critic_categories,user_score,user_categories,rating
0,Wii Sports,Wii,2006,sports,41.36,28.96,3.77,8.45,76,middle score,8,high score,E
2,Mario Kart Wii,Wii,2008,racing,15.68,12.76,3.79,3.29,82,high score,8,high score,E
3,Wii Sports Resort,Wii,2009,sports,15.61,10.93,3.28,2.95,80,high score,8,high score,E
6,New Super Mario Bros.,DS,2006,platform,11.28,9.14,6.50,2.88,89,high score,8,high score,E
7,Wii Play,Wii,2006,misc,13.96,9.18,2.93,2.84,58,middle score,6,middle score,E


- После категоризации данных проверьте результат: сгруппируйте данные по выделенным категориям и посчитайте количество игр в каждой категории.

In [46]:
# Проверим результат категоризации оценок критиков
df_actual.groupby('critic_categories')['name'].count()

critic_categories
high score      1686
middle score    5421
not rated       5665
Name: name, dtype: int64

In [47]:
# Проверим результат категоризации оценок пользователей
df_actual.groupby('user_categories')['name'].count()

user_categories
high score      2282
low score        116
middle score    4078
not rated       6296
Name: name, dtype: int64

- Выделим топ-7 платформ по количеству игр, выпущенных за весь актуальный период.

In [48]:
# Выводим топ-7 платформ по количеству игр, выпущенных за весь актуальный период
df_actual['platform'].value_counts().head(7)

PS2     2126
DS      2117
Wii     1275
PSP     1179
X360    1118
PS3     1087
GBA      810
Name: platform, dtype: int64

С небольшим отрывом первое место занимает PS2 с 2126 играми. За ней следует DS с 2117 играми. На третьем месте расположилась Wii с 1275 вышедшими играми.

---
<a class='anchor' id="link8"></a>
## 5. Итоговый вывод


В ходе предобработки данных мы проделали следующие действия:

- Ознакомились с датасетом, изучили его объем и форматы его данных

- Преобразовали названия столбцов и преобразовали типы данных в подходящие форматы

- Провели работу с пропусками в соответствии с характером данных

- Преобразовали данные для выявления неявных дубликатов и в последствии удалили все явные дубликаты

- Отфильтровали данные по году выпуска для дальнейшего исследования

- Добавили в датасет столбцы с категоризацией оценок критиков и пользователей

- Выделили топ-7 платформ по количеству вышедших игр за рассматриваемый период

 



<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:


Артур, здорово что в конце расписаны все этапы работы. Это важно потому что когда проект захочет посмотреть будущий работодатель (или начальник), у него может не быть времени на подробный разбор кода. Вероятнее всего он бегло просмотрит код, а из общего вывода захочет получить представление о всей работе.

 


<div class="alert alert-info">
<font size="5"><b>Комментарий ревьюера</b></font>


Артур, у тебя старательно выполненная работа, все четко, осмысленно. 
    
    
Выводы (в том числе промежуточные) присутствуют, они четкие и подробные 
    
    
Нет проблем с комментированием кода
    
     
Работа очень аккуратная, видно что приложены необходимо усилия
      
 
    
Советы:
    
     
- Попробуй поработать со стилистикой названий столбцов в автоматическом режиме	
	
	
 
- Вместо функция + apply  для заполнения пропусков в группировке, лучше использовать transformer  
        
     
	
- Можно немножко сэкономить   памяти и кое где использовать int  вместо float 		
	
     



Обязательно к исправлению:

 
  
- Перед исправлением типа данных стоит вывести уникальные значения по столбцам где мы видим несоответствия,  и проанализировать увиденное    
    
     
    
- Заглушки тут вообще не   обязательны,  но если используешь, то не забудь убрать их перед анализом    
    
 
Жду исправлений Артур, для принятия проекта. Если какие то вопросы, то сразу спрашивай )  
 

<div class="alert alert-info">
<font size="5"><b>Комментарий студента</b></font>


Большое спасибо, что подсветил слабые места в моём проекте! Надеюсь доработки получились корректные.


<div class="alert alert-info">
<font size="5"><b>Комментарий ревьюераV2</b></font>

Спасибо за работу!    

    
Красное исправлено,   желтые советы использованы ы  (это все было по желанию), значит стремишься развиваться, а желание и интерес это главное. Надеюсь мои советы и вопросики были полезны и в копилочку знаний упало что то новое, а проект стал лучше, и симпатичней.

 
  
Отличная работа Артур. Желаю успехов в дальнейшей учебе!

 